tracing.go

```diff
@@ -84,53 +87,155 @@ func (s Span) ParentID() uint64 { return s.parentID }
+		span := &Span{}
+		span.initFromOpenTracing(sp)
+		return span
 	}
+	return &emptySpan
 }
 
-// TracingEnabled checks whether tracing is enabled for this context
-func (s Span) TracingEnabled() bool { return (s.flags & tracingFlagEnabled) == tracingFlagEnabled }
+// emptySpan is returned from CurrentSpan(ctx) when there is no OpenTracing
+// Span in ctx, to avoid returning nil.
+var emptySpan Span
 
-// NewChildSpan begins a new child span in the provided Context
-func (s Span) NewChildSpan() *Span {
-	childSpan := &Span{
-		traceID:  s.traceID,
-		parentID: s.spanID,
-		flags:    s.flags,
+// initFromOpenTracing initializes Span fields from an OpenTracing Span,
+// assuming the tracing implementation supports Zipkin-style span IDs.
+func (s *Span) initFromOpenTracing(span opentracing.Span) {
+	if err := span.Tracer().Inject(span.Context(), ZipkinSpanFormat, s); err != nil {
+		s.initRandom()
 	}
-	if s.spanID == 0 {
-		childSpan.spanID = childSpan.traceID
-	} else {
-		childSpan.spanID = uint64(traceRng.Int63())
+}
+
+// startOutboundSpan creates a new tracing span to represent the outbound RPC call.
+// If the context already contains a span, it will be used as a parent, otherwise
+// a new root span is created.
+//
+// If the tracer supports Zipkin-style trace IDs, then call.callReq.Tracing is
+// initialized with those IDs. Otherwise it is assigned random values.
+func (c *Connection) startOutboundSpan(ctx context.Context, serviceName, methodName string, call *OutboundCall, startTime time.Time) opentracing.Span {
+	var parent opentracing.SpanContext // ok to be nil
+	if s := opentracing.SpanFromContext(ctx); s != nil {
+		parent = s.Context()
 	}
-	return childSpan
+	span := c.Tracer().StartSpan(
+		methodName,
+		opentracing.ChildOf(parent),
+		opentracing.StartTime(startTime),
+	)
+	if isTracingDisabled(ctx) {
+		ext.SamplingPriority.Set(span, 0)
+	}
+	ext.SpanKindRPCClient.Set(span)
+	ext.PeerService.Set(span, serviceName)
+	setPeerHostPort(span, c.remotePeerInfo.HostPort)
+	span.SetTag("as", call.callReq.Headers[ArgScheme])
+	call.callReq.Tracing.initFromOpenTracing(span)
+	return span
 }
 
-func (s *Span) sampleRootSpan(sampleRate float64) {
-	// Sampling only affects root spans
-	if s.ParentID() != 0 {
-		return
+// InjectOutboundSpan retrieves OpenTracing Span from `response`, where it is stored
+// when the outbound call is initiated. The tracing API is used to serialize the span
+// into the application `headers`, which will propagate tracing context to the server.
+// Returns modified headers containing serialized tracing context.
+func InjectOutboundSpan(response *OutboundCallResponse, headers map[string]string) map[string]string {
+	span := response.span
+	if span == nil {
+		return headers
+	}
+	if headers == nil {
+		headers = make(map[string]string)
+	}
+	carrier := opentracing.TextMapCarrier(headers)
+	if err := span.Tracer().Inject(span.Context(), opentracing.TextMap, carrier); err == nil {
+		return headers
```

Comment:
If it happens once, it's likely to keep happening so logging might be too much. Should we try to do this once at creation time to see if it works, and log once if it doesn't?

Or maybe it makes more sense as metrics? I'm happy to put a TODO and defer this for another PR.